### PR TITLE
haskell: Add hoogle option to developPackage

### PR DESCRIPTION
**Pull request by expipiplus1**
Also add `composeExtensionss` to lib.

This seems to be a sensible way of getting `ghcWithHoogle` into the mix.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by expipiplus1** [comment]
name hoogle -> withHoogle for consistency with other makeShell

File: lib/fixed-points.nix
Comment (cdepillabout) [inline lib/fixed-points.nix:85]: A few things:

1. I'm not a huge fan of this naming.  It almost looks like a spell miss.  However, I don't really have a suggestion for a better name.
2. Could you add either some sort of example to the docs, or maybe a type signature?  It is hard to tell how it should be used just from the comment on the function.
3. I'd suggestion pulling this out into a separate commit, since you wouldn't want it to get removed if the hoogle thing was reverted for whatever reason.

File: lib/tests/misc.nix
Comment (cdepillabout) [inline lib/tests/misc.nix:90]: Thanks for adding tests!

File: pkgs/development/haskell-modules/make-package-set.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/make-package-set.nix:224]:
```suggestion
    #   , withHoogle : Defaulted
```

File: pkgs/development/haskell-modules/make-package-set.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/make-package-set.nix:232]: There doesn't actually appear to be an explanation here of what `returnShellEnv` does.  Could you add this as well?

File: pkgs/development/haskell-modules/make-package-set.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/make-package-set.nix:245]: Why is this line needed?

Also, I don't quite understand the `ghc = super.ghc // { withPackages = super.ghc.withHoogle; };` line.

What is the effect of doing these overrides?  Could this cause any problems in any way?

**Review by cdepillabout — CHANGES_REQUESTED** [review]
I left a few comments.

Also, it would be nice if you wrote some documentation for end users who want to use Hoogle.  Could you add a sentence or two for what an end user would be able to do by enabling the `withHoogle` option?  I'm guessing the idea is that they now get a `hoogle` in their path with a database of all dependencies?

It'd also be nice to have a test for `developPackage` (similar to the test for `shellFor`), but that's more of a wish-list type thing.  Not having a test shouldn't block this from being merged in.

**Comment by expipiplus1** [comment]
> It'd also be nice to have a test for `developPackage` (similar to the test for `shellFor`), but that's more of a wish-list type thing. Not having a test shouldn't block this from being merged in.

Such a test would involve import-from-derivation, is that permissible in nixpkgs?

**Comment by cdepillabout** [comment]
> Such a test would involve import-from-derivation, is that permissible in nixpkgs?

Ah, good call.  

I guess you could always just mark it `__noChroot` and we could use it as a test that we run manually, but it would be unfortunate that we couldn't have it run by Hydra.

File: lib/fixed-points.nix
Comment (expipiplus1) [reply lib/fixed-points.nix]: Done done done

File: pkgs/development/haskell-modules/make-package-set.nix
Comment (expipiplus1) [reply pkgs/development/haskell-modules/make-package-set.nix]: I actually wrote this years ago and I'm a little sure that it was necessary then. Nevertheless I discovered envFunc today and this is gone :)

**Comment by expipiplus1** [comment]
> > Such a test would involve import-from-derivation, is that permissible in nixpkgs?
> 
> Ah, good call.
> 
> I guess you could always just mark it `__noChroot` and we could use it as a test that we run manually, but it would be unfortunate that we couldn't have it run by Hydra.

Right, perhaps best for another PR

File: lib/tests/misc.nix
Comment (cdepillabout) [inline lib/tests/misc.nix:92]: I think these tests need to be changed to use `composeManyExtensions` now.

**Review by cdepillabout — APPROVED** [review]
Aside from the one remaining comment, I think this looks really good.

I'm going to leave this open for a few days, so that any else that got notified has a chance to comment on the change in `lib/`.

@expipiplus1 If there are no additional comments to this in a few days, please feel free to ping me and I will merge it in (assuming the test thing has been fixed).

**Comment by expipiplus1** [comment]
Will do, thanks

**Comment by expipiplus1** [comment]
@cdepillabout ping!

**Comment by expipiplus1** [comment]
I don't think that @ofborg's failure is to do with this PR

**Review by cdepillabout — APPROVED** [review]

**Comment by expipiplus1** [comment]
Thanks!

**Comment by FRidh** [comment]
`composeManyExtensions` was proposed in https://github.com/NixOS/nixpkgs/issues/33258.

**Pull request merged**
